### ISSUE TITLE
REST API "The product can't be saved"

### ISSUE THREAD

**Issue by Zyles** [magento/magento2#26053]
Another case of bad error message feedback.

No logs other than the REST API giving terrible error message and a whole bunch of Googling has to be done.

In my case I tried to upload an image through the rest API using POST to `/rest/V1/products/<sku>/media`

The issue has to do with products having required fields that are not filled in. This could arise after changing attribute sets and not going back to fix it on old products for example.

The solution is to go into admin, filling in the required fields for the product, then trying to upload the image again.

### Preconditions (*)

1. Magento 2.3.3

### Steps to reproduce (*)

1. Have required fields on product that are not filled in
1. Try to upload image using REST API

### Expected result (*)

1. Better error message that is actually helpful.
2. Log output in var/log

### Actual result (*)

1. Terrible error message.
2. No logs
3. Not being able to upload the product image

**Comment by engcom-Delta** [comment]
@Zyles , thank you for your report. 
Looks like it would be good solution to add output with stack trace in var/log. For now response is `"message": "The product can't be saved."` 
Marked as **Future Request**
cc @sdzhepa

**Comment by LiamKarlMitchell** [comment]
Related files.
```
vendor/magento/module-catalog/Model/Product/Gallery/GalleryManagement.php
79:            throw new StateException(__("The product can't be saved."));
129:            throw new StateException(__("The product can't be saved."));

vendor/magento/module-catalog/Test/Unit/Model/Product/Gallery/GalleryManagementTest.php
85:     * @expectedExceptionMessage The product can't be saved.
160:     * @expectedExceptionMessage The product can't be saved.
```

Change to include exception as 2nd argument to new StateException so that it is actually helpful.
```php
 throw new StateException(__("The product can't be saved."), $e);
```

Also can throw for other reasons such as "The image MIME type is not valid or not supported." (Corrupted image or wrong extension/mimetype?)

**Comment by LiamKarlMitchell** [comment]
```php
try {
            $product = $this->productRepository->save($product);
        } catch (InputException $inputException) {
            throw $inputException;
        } catch (\Exception $e) {
            throw new StateException(__('Cannot save product. [%1]', $e->getMessage()));
        }
```
Responds with
```json
{
    "message": "Cannot save product. [%1]",
    "parameters": [
        "The image does not exist."
    ]
}
```
So then the client using the API has to put the parameters into the message.

Why "The image does not exist." I want to upload an image. (Magento 2.2.0)
The image is base64 encoded.

Another inner exception I got when I changed the filename was "Unable to save product"

In this case that I ran into,
When saving from Magento admin turns out permissions are denied on the cache folder...

Unable to write file into directory /var/www/html/pub/media/catalog/product/cache/7690288cef886e3600a2dbabc56d79c1/t/e. Access forbidden.

Ownership of the directory was wrong after migrating with rsyncing the folder to another server.

It does look like var/log/exception.log would have told me this if I had looked, so long as the critical/warning inner exceptions get logged I suppose that is fine, but the wording on the error could indicate a bit more as to why. E.g. "Permission error check exception log for more details."
```main.CRITICAL: Warning: rename(/var/www/html/pub/media/tmp/catalog/product/T/1/Test-1.jpg): failed to open stream: Permission denied in /var/www/html/vendor/magento/framework/File/Uploader.php on line 283 ```

**Comment by MJTheOne** [comment]
Why has this gone stale? This is a very annoying bump I'm also hitting. My import worked fine 5 minutes ago, added a custom_attribute to determine the url_key because I'm import products with same names. And boom an error which I can't debug because all it tells me is: "Product is unable to be saved"...

**Comment by dverkade** [comment]
Posting a comment. Issues should not be automatically closed.

**Comment by LudwigDev9** [comment]
Any update to the issue?

**Comment by stereoscope** [comment]
hello. the same issue occurred just yesterday. while tests where running against Magento v2.4.4. 
i get the same error .... 

here my docker-compose script

version: '2'
services:
  mariadb:
    image: docker.io/bitnami/mariadb:10.3
    environment:
      # ALLOW_EMPTY_PASSWORD is recommended only for development.
      - ALLOW_EMPTY_PASSWORD=yes
      - MARIADB_USER=bn_magento
      - MARIADB_DATABASE=bitnami_magento
    volumes:
      - 'mariadb_data:/bitnami/mariadb'
  magento:
    image: docker.io/bitnami/magento:2.4.4
    ports:
      - '80:8080'
      - '443:8443'
    environment:
      - MAGENTO_HOST=localhost
      - MAGENTO_DATABASE_HOST=mariadb
      - MAGENTO_DATABASE_PORT_NUMBER=3306
      - MAGENTO_DATABASE_USER=bn_magento
      - MAGENTO_DATABASE_NAME=bitnami_magento
      - ELASTICSEARCH_HOST=elasticsearch
      - ELASTICSEARCH_PORT_NUMBER=9200
      # ALLOW_EMPTY_PASSWORD is recommended only for development.
      - ALLOW_EMPTY_PASSWORD=yes
    volumes:
      - 'magento_data:/bitnami/magento'
    depends_on:
      - mariadb
      - elasticsearch
  elasticsearch:
    image: docker.io/bitnami/elasticsearch:7.16.3
    volumes:
      - 'elasticsearch_data:/bitnami/elasticsearch/data'
volumes:
  mariadb_data:
    driver: local
  magento_data:
    driver: local
  elasticsearch_data:
    driver: local

**Comment by TysvdHeuvel** [comment]
> 

Thanks Ludwig! 2 hours of research.. but thanks to this hint I managed to solve my problem. 
How did you find out? My logs don't show anything that points in this direction.

**Comment by ViniciusKoiti** [comment]
I have same problem problem on  update, Magento\\Catalog\\Model\\Product\\Gallery\\GalleryManagement->update('42489', Object(Magento\\Catalog\\Model\\Product\\... (1988 bytes)], someone now anything about?

**Comment by ViniciusKoiti** [comment]
> 

How I try to send images, I give that sometimes, but other times I gave success in same object

**Comment by yhling** [comment]
If you're using a Bitnami docker image, make sure `/opt/bitnami/php/tmp` is writeable by PHP

Run this inside the magento container terminal:

`chmod 777 /opt/bitnami/php/tmp`

Doing this solved the "The product can't be saved" for me

**Comment by woutertje99** [comment]
I am doing a batch image import (one by one) and for me, it seems to be the JPG extension. As far as I can see, my PNG files all seem to be working, but the JPG give errors. All of them are uploaded in Base64.
I'll check and comment back if I manage to find a solution!

**Comment by woutertje99** [comment]
> I am doing a batch image import (one by one) and for me, it seems to be the JPG extension. As far as I can see, my PNG files all seem to be working, but the JPG give errors. All of them are uploaded in Base64. I'll check and comment back if I manage to find a solution!

Nope, that is not the issue. Second guess is corrupted Base64, or missing data in the import.
Magento is such a pain in the ass to work with, no error logs, no support given anymore, not the best documentation...
